diff --git a/nixos/modules/services/networking/nftables.nix b/nixos/modules/services/networking/nftables.nix
index fc823b5ff597..b017564af195 100644
--- a/nixos/modules/services/networking/nftables.nix
+++ b/nixos/modules/services/networking/nftables.nix
@@ -323,8 +323,6 @@ in
             executable = true;
             text = ''
               #! ${pkgs.nftables}/bin/nft -f
-              # previous deletions, if any
-              include "${deletionsScriptVar}"
               # current deletions
               include "${deletionsScript}"
               ${concatStringsSep "\n" (
